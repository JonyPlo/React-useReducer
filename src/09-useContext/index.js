export * from './MainApp';
export * from './HomePage';
export * from './AboutPage';
export * from './LoginPage';
export * from './ErrorPage';
export * from './Navbar';
