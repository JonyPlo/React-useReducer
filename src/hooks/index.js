export * from './useCounter'
export * from './useFetch'
export * from './useFetchAllQuotes'
export * from './useForm'