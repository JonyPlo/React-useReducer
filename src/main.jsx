import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import {
  createBrowserRouter,
  RouterProvider,
  Navigate,
} from 'react-router-dom';
import {
  MainApp,
  HomePage,
  AboutPage,
  LoginPage,
  ErrorPage,
} from './09-useContext';
// import { CounterApp } from './01-useState/CounterApp'
// import { HooksApp } from './HooksApp'
// import { CounterWithCustomHook } from './01-useState/CounterWithCustomHook'
// import { SimpleForm } from './02-useEffect/SimpleForm';
// import { FormWithCustomHook } from './02-useEffect/FormWithCustomHook';
// import { MultipleCustomHooks } from './03-examples/MultipleCustomHooks';
// import { FocusScreen } from './04-useRef/FocusScreen';
// import { Layout } from './05-useLayoutEffect/Layout';
// import { Memorize } from './05-memos/Memorize';
// import { MemoHook } from './05-memos/MemoHook';
// import { CallbackHoock } from './05-memos/CallbackHoock';
// import { Padre } from './07-tarea-memo/Padre';
// import './08-useReducer/intro-reducer';
// import { TodoApp } from './08-useReducer/TodoApp';

const router = createBrowserRouter([
  {
    path: '/',
    element: <MainApp />,
    errorElement: <Navigate to={'about'} />,
    children: [
      {
        path: '/',
        element: <HomePage />,
      },
      {
        path: 'about',
        element: <AboutPage />,
      },
      {
        path: 'login',
        element: <LoginPage />,
      },
    ],
  },
]);

ReactDOM.createRoot(document.getElementById('root')).render(
  //  <React.StrictMode>
  <RouterProvider router={router} />
  //  </React.StrictMode>
);
