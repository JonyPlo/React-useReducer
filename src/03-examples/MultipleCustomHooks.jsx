import { useCounter, useFetch, useFetchAllQuotes } from '../hooks';
import { LoadingQuote, Quote } from './';

export const MultipleCustomHooks = () => {
  const { counter, increment, decrement } = useCounter(1);
  const { data, isLoading, hasError } = useFetch(
    `http://localhost:3000/breakingbad?quote_id=${counter}`
  );
  const { allQuotes } = useFetchAllQuotes('http://localhost:3000/breakingbad');
  const { quote, author } = !!data && data[0]; //Aqui aplicamos doble negacion (!!) se hace esto porque "data" al principio es null por lo tanto no se puede desestructurar algo de null, asi que al aplicar la primera negacion, osea !data, transformamos el valor de null a true, y con la segunda negacion !!data convertimos el valor de true a false, por lo tanto de esta forma estamos haciendo que solo se desestructure cuando data tenga datos.
  // Otra opcion para convertir el data en un Booleano seria con la funcion de javascript Boolean(data), esto convertira al data con el valor null en un booleano, en este caso como el valor es null entonces el Boolean() lo convertira en false hasta que tenga datos y lo convertira a true

  const disabledButtonNext = isLoading || counter === allQuotes.length;
  const disabledButtonPrevious = isLoading || counter === 1;

  return (
    <>
      <h1>BreakingBad Quotes</h1>
      <hr />

      {isLoading ? <LoadingQuote /> : <Quote quote={quote} author={author} />}

      <button
        className='btn btn-primary'
        disabled={disabledButtonPrevious}
        onClick={() => decrement()}
      >
        Prev Quote
      </button>
      <button
        className='btn btn-primary'
        disabled={disabledButtonNext}
        onClick={() => increment()}
      >
        Next Quote
      </button>
    </>
  );
};
