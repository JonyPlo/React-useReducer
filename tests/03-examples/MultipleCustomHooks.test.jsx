import { fireEvent, render, screen } from '@testing-library/react';
import { MultipleCustomHooks } from '../../src/03-examples';
import { useFetch } from '../../src/hooks/useFetch';
import { useCounter } from '../../src/hooks/useCounter';

jest.mock('../../src/hooks/useFetch'); //Tener en cuenta que cuando se hace un mock, no se debe apuntar al archivo de barril porque si el componente usa otro custom hook, tambien lo convertira en mock y eso nos dara error, lo mejor es apuntar directamente al hook
jest.mock('../../src/hooks/useCounter');

describe('Pruebas en MultipleCustomHooks', () => {
  const mockIncrement = jest.fn();

  useCounter.mockReturnValue({
    counter: 1,
    increment: mockIncrement,
  });

  //Para asegurarme de que todas las funciones a las que les estoy haciendo un mock sean reseteadas, usamos una parte del ciclo de vida de las pruebas que es "beforeEach", este significa "antes de cada prueba", y con jest.clearAllMocks() limpiamos todos los mock y asi en caso que un mock ya haya sido llamado por un test, al ejecutar el test siguiente que use el mismo mock lo usara con los valores reseteados
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Debe de hacer match con el snapshot', () => {
    //! Si el componente que renderizamos llama a un hook que previamente convertimos en un mock y a ese mock lo necesitamos con valores de retorno diferentes en cada test, hay que llamar al mock en todos los tests antes de renderizar el componente, por ejemplo en este caso, llamar al useFetch.mockReturnValue() antes del render del componente <MultipleCustomHooks />. Pero si el mock que queremos usar tendra siempre los mismos valores de retorno, entonces a este mock podemos definirlo en el describe fuera de los test, como se hizo mas arriba con el mock de useCounter() para que esto afecte a todos los tests
    useFetch.mockReturnValue({
      data: null,
      isLoading: true,
      hasError: null,
    });

    const { container } = render(<MultipleCustomHooks />);
    expect(container).toMatchSnapshot();
  });

  test('Debe de mostrar el componente por defecto ', () => {
    useFetch.mockReturnValue({
      data: null,
      isLoading: true,
      hasError: null,
    });

    render(<MultipleCustomHooks />);

    expect(screen.getByText('Loading...'));
    expect(screen.getByText('BreakingBad Quotes'));

    const nextButton = screen.getByRole('button', { name: 'Next Quote' });
    const prevButton = screen.getByRole('button', { name: 'Prev Quote' });

    expect(nextButton.disabled).toBeTruthy();
    expect(prevButton.disabled).toBeTruthy();
  });

  test('Debe de mostrar un Quote ', () => {
    useFetch.mockReturnValue({
      data: [
        {
          quote: 'Hola Mundo',
          author: 'Jony',
        },
      ],
      isLoading: false,
      hasError: null,
    });

    render(<MultipleCustomHooks />);

    expect(screen.getByText('Hola Mundo')).toBeTruthy();
    expect(screen.getByText('Jony')).toBeTruthy();

    const nextButton = screen.getByRole('button', { name: 'Next Quote' });
    const prevButton = screen.getByRole('button', { name: 'Prev Quote' });

    expect(nextButton.disabled).toBeFalsy();
    expect(prevButton.disabled).toBeTruthy();
  });

  test('Debe de llamar a la funcion de incrementar ', () => {
    useFetch.mockReturnValue({
      data: [
        {
          quote: 'Hola Mundo',
          author: 'Jony',
        },
      ],
      isLoading: false,
      hasError: null,
    });

    render(<MultipleCustomHooks />);

    const nextButton = screen.getByRole('button', { name: 'Next Quote' });

    fireEvent.click(nextButton);

    expect(mockIncrement).toHaveBeenCalled();
  });
});
