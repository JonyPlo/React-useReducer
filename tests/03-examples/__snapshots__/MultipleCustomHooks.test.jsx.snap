// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    class="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    class="btn btn-primary"
    disabled=""
  >
    Prev Quote
  </button>
  <button
    class="btn btn-primary"
    disabled=""
  >
    Next Quote
  </button>
</div>
`;
